### PR TITLE
#1187 using options during elections

### PR DESCRIPTION
**Pull request by krzyk**
#1187 
* using options maxJobsInAgenda during elections (using new voter)
* refactored test elections_dont_exceed_max_jobs_option

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1446](https://github.com/zerocracy/farm/issues/1446) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1446](https://github.com/zerocracy/farm/issues/1446) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2001646332); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobs.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    "%d job(s) already, maxJobsInAgenda option is %d"
+                ).say(total, max)
+            );
+        } else {
+            rate = 0.0d;
+            log.append(
+                new Par(
+                    "%d job(s) out of %d from maxJobsInAgenda option"
+                ).say(total, max)
+            );
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobs.java]: @krzyk Is ``Par.say(...)`` necessary here? Can't we just use ``String.format(...)``?

File: src/main/java/com/zerocracy/pmo/Options.java [hunk 2]
```diff
@@ -61,23 +62,22 @@ public Options(final Pmo pmo, final String login) {
      * @throws IOException If fails
      */
     public Options bootstrap() throws IOException {
-        try (final Item team = this.item()) {
-            new Xocument(team).bootstrap("pmo/options");
+        try (final Item itm = this.item()) {
+            new Xocument(itm).bootstrap("pmo/options");
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/pmo/Options.java]: @krzyk Can you leave the ``team`` variable or whatever it is? ``itm`` is generic, what is it actually?

File: src/main/java/com/zerocracy/pmo/Options.java [hunk 3]
```diff
@@ -61,23 +62,22 @@ public Options(final Pmo pmo, final String login) {
...
      * @return Jobs number
      * @throws IOException If fails
      */
-    public int maxJobsInAgenda(final int def) throws IOException {
+    public int maxJobsInAgenda() throws IOException {
         try (final Item item = this.item()) {
             return new IoCheckedScalar<>(
                 new ItemAt<Number>(
-                    xpath -> def,
+                    xpath -> Integer.MAX_VALUE,
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/pmo/Options.java]: @krzyk Is there a situation where a user can have unlimited number of tasks in agenda, or?

File: src/main/java/com/zerocracy/pmo/Options.java [hunk 4]
```diff
@@ -88,34 +88,47 @@ public int maxJobsInAgenda(final int def) throws IOException {
...
     /**
      * Notify publish option.
-     * @param def Default value
      * @return True if set
      * @throws IOException If fails
      */
-    public boolean notifyPublish(final boolean def) throws IOException {
-        return this.notify("publish", def);
+    public boolean notifyPublish() throws IOException {
+        return this.notify("publish", true);
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/pmo/Options.java]: @krzyk Why these changes related to notifying students? They don't seem part of the task.

File: src/test/resources/com/zerocracy/bundles/elections_dont_exceed_max_jobs_option/_after.groovy [hunk 5]
```diff
@@ -17,31 +17,22 @@
...
-//    new Footprint(farm, project).collection().find(
-//      Filters.and(
-//        Filters.eq('project', project.pid()),
-//        Filters.eq('type', 'Performer was elected')
-//      )
-//    )
-//  ).value() == 0
+  Farm farm = binding.variables.farm
+  assert new Agenda(farm, 'yegor256').bootstrap().jobs().size() == 1
+  assert new LengthOf(
```
Comment (amihaiemil) [inline src/test/resources/com/zerocracy/bundles/elections_dont_exceed_max_jobs_option/_after.groovy]: @krzyk Does ``assert`` have something to offer more than ``MatcherAssert``? I think ``MA`` is used in other groovy tests as well.

File: src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy [hunk 6]
```diff
@@ -93,6 +93,7 @@ def exec(Project project, XML xml) {
         (new VsSafe(new VsRate(project, logins)))                                 : 2,
         (new VsSafe(new VsBigDebt(pmo)))                                          : -100,
         (new VsSafe(new VsNoRoom(pmo)))                                           : role == 'REV' ? 0 : -100,
+        (new VsSafe(new VsOptionsMaxJobs(pmo)))                                   : role == 'REV' ? 0 : -100,
```
Comment (amihaiemil) [inline src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy]: @krzyk What's the difference between ``VsOptionsMaxJobs`` and ``VsNoRoom``?

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @since 1.0
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class VsOptionsMaxJobsTest {
+
+    @Test
+    public void votesHighIfMaxJobsReached() throws Exception {
+        final Project project = new FkProject();
+        final FkFarm farm = new FkFarm();
```
Comment (amihaiemil) [inline src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @krzyk I see ``farm`` and ``project`` are used only once, can we inline them?

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        final String user = "g4s8";
+        final int total = 10;
+        final Pmo pmo = new Pmo(farm);
+        new Options(pmo, user).bootstrap().maxJobsInAgenda(total);
+        final Agenda agenda = new Agenda(pmo, user).bootstrap();
+        for (int num = 0; num < total; ++num) {
+            agenda.add(project, String.format("gh:test/test#%d", num), "DEV");
```
Comment (amihaiemil) [inline src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @krzyk Same with ``agenda``

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void votesLowHasRoom() throws Exception {
+        final String user = "krzyk";
+        final Pmo pmo = new Pmo(new FkFarm());
+        new Options(pmo, user).bootstrap().maxJobsInAgenda(1);
+        MatcherAssert.assertThat(
+            new VsOptionsMaxJobs(pmo).take(user, new StringBuilder(0)),
+            // @checkstyle MagicNumberCheck (1 lines)
+            Matchers.closeTo(0.0, 0.001)
```
Comment (amihaiemil) [inline src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @krzyk I remember there was a reason for using ``closeTo`` in another PR, but in the case of this voter, couldn't we simply assert for equality to ``1`` in the unit test above and ``0`` here?

File: src/main/java/com/zerocracy/pmo/Options.java [hunk 10]
```diff
@@ -88,34 +88,47 @@ public int maxJobsInAgenda(final int def) throws IOException {
...
+    public void maxJobsInAgenda(final int max) throws IOException {
+        try (final Item item = this.item()) {
+            new Xocument(item.path()).modify(
+                new Directives()
+                    .xpath("/options")
+                    .addIf("maxJobsInAgenda")
+                    .set(max)
+            );
+        }
+    }
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/pmo/Options.java]: @krzyk Is the change to these ``maxJobsInAgenda`` part of the task?

File: src/main/java/com/zerocracy/pmo/Options.java (hunk 2 above)
Comment (krzyk) [reply src/main/java/com/zerocracy/pmo/Options.java]: @amihaiemil well, definatelly not a `team` :)
And item is just the path to the options file in this case, I changed it because `team` was misleading

File: src/main/java/com/zerocracy/pmo/Options.java (hunk 3 above)
Comment (krzyk) [reply src/main/java/com/zerocracy/pmo/Options.java]: @amihaiemil I don't get what is the point of question here :)
If user can have unlimited, I don't know maybe, MAX_VALUE is pretty close to unlimited. If user doesn't set this option we can safely assume that he has no trouble with accepting MAX_VALUE items :)
The limiting is either way applied later on during the check of number of points.

File: src/main/java/com/zerocracy/pmo/Options.java (hunk 10 above)
Comment (krzyk) [reply src/main/java/com/zerocracy/pmo/Options.java]: @amihaiemil yes, it is needed in tests

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java (hunk 8 above)
Comment (krzyk) [reply src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @amihaiemil no, because it will be needlesly created `total` number of times (and even further slow down the tests)

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java (hunk 9 above)
Comment (krzyk) [reply src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @amihaiemil well, you can't compare doubles/floats directly (https://stackoverflow.com/questions/8081827/how-to-compare-two-double-values-in-java)

File: src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy (hunk 6 above)
Comment (krzyk) [reply src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy]: @amihaiemil VsNoRoom uses points to check if person has a room in agenda based on points http://www.zerocracy.com/policy.html#3

VsOptionsMaxJobs - furhter allows limiting of the jobs below the values from VsNoRoom

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java (hunk 7 above)
Comment (krzyk) [reply src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @amihaiemil project is used in a loop, creating new FkProject will result in error (because agenda will have jobs from different projects)

**Comment by krzyk** [comment]
@amihaiemil updated the code, and replied to some comments

File: src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java (hunk 8 above)
Comment (amihaiemil) [reply src/test/java/com/zerocracy/pm/staff/votes/VsOptionsMaxJobsTest.java]: @krzyk Right, my mistake, sorry

**Comment by amihaiemil** [comment]
@rultor merge

File: src/test/resources/com/zerocracy/bundles/elections_dont_exceed_max_jobs_option/_after.groovy [hunk 11]
```diff
@@ -17,31 +17,30 @@
...
+  MatcherAssert.assertThat(
+    new LengthOf(
+      new Footprint(farm, project).collection().find(
+        Filters.and(
+          Filters.eq('project', project.pid()),
+          Filters.eq('type', 'Performer was elected')
+        )
+      )
+    ).value().intValue(),
+    Matchers.is(0)
```
Comment (g4s8) [inline src/test/resources/com/zerocracy/bundles/elections_dont_exceed_max_jobs_option/_after.groovy]: @krzyk maybe it's better to use `Matchers.emptyIterable()` or `Matchers.empty()` here?

**Comment by krzyk** [comment]
@g4s8 corrected, take a look

**Comment by g4s8** [comment]
@krzyk thanks

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1446](https://github.com/zerocracy/farm/issues/1446) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +31 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)
